keps/2941-DRA/README.md

```diff
@@ -0,0 +1,729 @@
+
+In order to mitigate this risk, Kueue can take the following approach:
+1. For DRAPrioritizedLists: all the mentioned device classes in the request will be counted against the quota
+2. For DRAAdminAccess: This feature can only be enabled in admin namespace, therefore it should be skipped for being 
+   counted against ClusterQuota. This is a different stance than Kubernetes ResourceQuota because kubernetes
+   ResourceQuota is namespace scoped. As a result, admin users can account for quota independent of user workloads.
+   On the contrary, with Kueue, since quota is part of ClusterQuota a cluster scoped object, admin workloads using
+   devices in admin namespace if counted against quota, will eat up quota meant for user workloads.
+3. For ResourceClaims with allocation mode `All`: worst-case scenario of the max number of devices that could be
+   allocated to a single claim will be used against quota.
+
+
+## Design Details
+
+A new feature gate DynamicResourceAllocation will be introduced, allowing users to test it in dev environments while
+making the changes dormant for production users. The following sections will explain the design in detail.
+
+### DeviceClassMapping API
+
+```golang
+// DeviceClassMapping is a cluster-scoped singleton CRD that maps device classes to canonical names
+// used in ClusterQueue resource quotas. It is singleton as "default" is the only allowed named for the CRD instance.
+type DeviceClassMapping struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec defines the desired state of DeviceClassMapping
+    Spec DeviceClassMappingSpec `json:"spec"`
+}
+
+// DeviceClassMappingSpec defines the mappings between device classes and canonical names
+type DeviceClassMappingSpec struct {
+    // DynamicResources is a list of mappings from canonical names to device classes
+    // +listType=atomic
+    DynamicResources []DeviceMapping `json:"dynamicResources"`
+}
+
+// DeviceMapping defines a mapping from a canonical name to a list of device classes
+type DeviceMapping struct {
+    // Name is the canonical name that will be referred to in ClusterQueue and Workload status
+    Name corev1.ResourceName `json:"name"`
+
+    // DeviceClassNames lists the names of all the device classes that will count against
+    // the quota defined for this canonical name
+    // +listType=atomic
+    DeviceClassNames []corev1.ResourceName `json:"deviceClassNames"`
+}
+```
+
+The ResourceFlavor spec will have a field called dynamicResource. The cluster admin has to define the list of
+deviceClasses that will be represented by a canonical name and can be used to define quotas in cluster queue. With this
+a ClusterQueue with nominalQuota can be defined as follows:
+
+```yaml
+apiVersion: kueue.x-k8s.io/v1alpha1
+kind: DeviceClassMapping
+metadata:
+  name: "default"  # Fixed name - singleton CR, only one instance allowed
```

Comment:
I would probably suggest keeping it under kueue-system namespace. This way it is closer to the ConfigMap model, which is intrinsically namespaced. Maybe using helm one could imagine two separete installations of Kueue. It is very unlikely, but possible.

Reply:
As I posted in https://github.com/kubernetes-sigs/kueue/pull/5149#discussion_r2161830900, I'm leaning toward to (3). The cluster admin can select which cq can manage which namespaces based on namespaceSelector in cq.

After we introduce TAS to kueue, kueue-controller-manager retrieves all node information as a dedicated cache separate from the informer cache, which increases the memory footprint and kueue scheduling duration.

The sharding kueue-controller-manager concept would mitigate this problem.